src/main/java/io/nats/client/impl/NatsConnectionWriter.java

```diff
@@ -116,15 +116,13 @@ synchronized void sendMessageBatch(NatsMessage msg, DataPort dataPort, NatsStati
             if (sendPosition + size > sendBuffer.length) {
                 if (sendPosition == 0) { // have to resize
                     this.sendBuffer = new byte[(int)Math.max(sendBuffer.length + size, sendBuffer.length * 2L)];
-                } else { // else send and go to next message
+                } else { // else send and continue with current message
                     dataPort.write(sendBuffer, sendPosition);
                     connection.getNatsStatistics().registerWrite(sendPosition);
                     sendPosition = 0;
-                    msg = msg.next;
 
-                    if (msg == null) {
-                        break;
-                    }
+                    // go back to the start of the loop, to ensure we check for resizing the buffer again
+                    continue;

```

Comment:
The `continue` moves to the top of the loop, skipping the end of the loop. Maybe 
```
msg = msg.next;
continue;
```